### PR TITLE
Add code for speaker embeddings task

### LINKED ISSUE #25: Task: Data representation
### Description

Extracting audio self-supervised models' embeddings, supervised model's speaker embeddings (xvector, ECAPA-TDNN, ResNet), acoustic features, linguistic features, and text self-supervised model's embeddings.

### Tasks

- [x] Extracting audio self-supervised models' embeddings @GasserElbanna , 
- [x] Extracting text self-supervised model's embeddings,
- [x] Extracting supervised model's speaker embeddings (xvector, ECAPA-TDNN, ResNet) @900miles , 
- [ ] Extracting handcrafted features (opensmile, praat, torchaudio - including spectrogram - and consider integrating features from [Nick’s table](https://docs.google.com/spreadsheets/d/18V_FrE3jYm1Msl4rg8xl8WVBDprbZrRFIVm3nEf2uTQ/edit?usp=sharing) and [Daniel's paper](https://onlinelibrary.wiley.com/doi/pdf/10.1002/lio2.354) in the future),
- [ ] Extracting linguistic features.

### Freeform Notes

_No response_

### LINKED ISSUE #73: Extract supervised model's speaker embeddings
### Description

Description:
Implement functionality to extract speaker embeddings from a given tensor using any of xvector, ECAPA-TDNN, or ResNet.

### Tasks

- [x] xvector
- [x] ECAPA-TDNN
- [x] ResNet

### Freeform Notes

From #25

### PR DESCRIPTION
**Pull request by 900miles**
This fails some pre-commit tests on openmind, not sure why.

### CODE REVIEW

**Comment by fabiocat93** [comment]
> This fails some pre-commit tests on openmind, not sure why.

I have fixed that by adding requests as a required dependency to the `mypy` pre-commit hook.

**Comment by fabiocat93** [comment]
@900miles i have noticed that speechbrain creates a folder for checkpoints called `pretrained_models`. 2 things:
- do you know if there is a way to customize the directory?
- can you please add the default directory to the .gitignore file? 
thank you!

File: src/senselab/audio/tasks/speaker_embeddings/api.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+import pydra
+from speechbrain.inference.speaker import EncoderClassifier
+from torch import Tensor
+
+from senselab.audio.data_structures.audio import Audio
+from senselab.utils.data_structures.device import DeviceType
+from senselab.utils.data_structures.model import HFModel, SenselabModel
+
+
+def extract_embeddings(audio: Audio, model: SenselabModel, device: Optional[DeviceType] = None) -> Tensor:
```
Comment (fabiocat93) [inline src/senselab/audio/tasks/speaker_embeddings/api.py]: - can you please make this `extract_speaker_embeddings_from_audios`?
- can you please make this work with a list of audios?

File: src/senselab/audio/tasks/speaker_embeddings/api.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    Returns:
+        Tensor: A tensor containing the speaker embeddings.
+
+    Raises:
+        NotImplementedError: If the model is not a Hugging Face model.
+
+    Examples:
+        >>> audio = Audio.from_filepath("sample.wav")
+        >>> model = HFModel(path_or_uri="speechbrain/spkrec-ecapa-voxceleb")
+        >>> embeddings = extract_embeddings(audio, model, device="cuda")
```
Comment (fabiocat93) [inline src/senselab/audio/tasks/speaker_embeddings/api.py]: device is of type DeviceType and is not a string as it seems from your examples.

File: src/senselab/audio/tasks/speaker_embeddings/api.py (hunk 1 above)
Comment (fabiocat93) [reply src/senselab/audio/tasks/speaker_embeddings/api.py]: - are CPU, CUDA, and MPS supported? if not, please use the same logic as in the task `speech_to_text` to handle this

Comment (fabiocat93) [reply src/senselab/audio/tasks/speaker_embeddings/api.py]: - i know that for now all models come from speechbrain. however, in the future we may want to integrate models from other libraries (e.g., `espnet`, `pyannote.audio`). hence, can you please create a python file specific for speech brain and make the API method here a wrapper of that method? same as in the `speech_to_text` task. thank you!

Comment (fabiocat93) [reply src/senselab/audio/tasks/speaker_embeddings/api.py]: - last, can you add in the documentation of the speechbrain function that the default threshold for determining if two audios are from the same or different speakers is 0.25 [ref: https://speechbrain.readthedocs.io/en/latest/API/speechbrain.inference.speaker.html#speechbrain.inference.speaker.SpeakerRecognition.verify_batch]

Comment (900miles) [reply src/senselab/audio/tasks/speaker_embeddings/api.py]: Where should I add this? Would it not be better to put that in the speaker verification documentation?

**Comment by 900miles** [comment]
All of the suggested changes have been pushed.

> @900miles i have noticed that speechbrain creates a folder for checkpoints called `pretrained_models`. 2 things:
> 
>     * do you know if there is a way to customize the directory?
> 
>     * can you please add the default directory to the .gitignore file?
>       thank you!

I added the default directories to the .gitignore file. I'm not sure if there's a way to customize the directory, and a quick google wasn't very helpful either.

File: src/senselab/audio/tasks/speaker_embeddings/api.py (hunk 1 above)
Comment (fabiocat93) [reply src/senselab/audio/tasks/speaker_embeddings/api.py]: Currently, we don't have speaker verification as a task. Speaker verification is generally performed by computing cosine similarity between speaker embeddings. I would add the reference here for now but agree with you that we may want to create a workflow/task for speaker verification.

**Comment by fabiocat93** [comment]
> All of the suggested changes have been pushed.
> 
> > @900miles i have noticed that speechbrain creates a folder for checkpoints called `pretrained_models`. 2 things:
> > ```
> > * do you know if there is a way to customize the directory?
> > 
> > * can you please add the default directory to the .gitignore file?
> >   thank you!
> > ```
> 
> I added the default directories to the .gitignore file. I'm not sure if there's a way to customize the directory, and a quick google wasn't very helpful either.

```
classifier = EncoderClassifier.from_hparams(
    ...     source="speechbrain/spkrec-ecapa-voxceleb",
    ...     savedir=tmpdir,
    ... )
```

**Comment by fabiocat93** [comment]
Thank you @900miles ! I have added a couple of edits (mostly for making the SpeechBrain class a factory). there are a couple of todoes, but they are minor and we can keep them for later:
- we may want to have a standard repo for saving pre-trained models
- we may want to better check if a model is a speechbrain model (being a hf model is not enough)

**Pull request merged**